src/main/java/build/buildfarm/server/ByteStreamService.java

```diff
@@ -262,9 +262,11 @@ private void writeOperationStream(
         String operationStream = parseOperationStream(resourceName);
 
         OutputStream outputStream = instance.getStreamOutput(operationStream);
-        outputStream.write(request.getData().toByteArray());
-        if (request.getFinishWrite()) {
-          outputStream.close();
+        if (outputStream != null) {
+          request.getData().writeTo(outputStream);
+          if (request.getFinishWrite()) {
+            outputStream.close();
```

Comment:
I'm open to suggestions - it should be the instance's responsibility to maintain that strategy, and making a forced close at complete operation post or requeue seems like a start.